src/persist-txn/src/txn_cache.rs

```diff
@@ -1049,127 +1049,123 @@ mod tests {
-        cache.push_append(other, vec![0xB0], 7, 1);
-        cache.push_append(d0, vec![0xA0], 7, -1);
-        cache.progress_exclusive = 8;
-        assert_eq!(cache.data_listen_next(&d0, 7), EmitLogicalProgress(8));
-
-        // Unlike a snapshot, the previous answers have changed! This is because
-        // it's silly (inefficient) to walk through the state machine for a big
-        // range of time once you know of a later write that advances the
-        // physical upper.
-        // assert_eq!(cache.data_listen_next(&d0, 3), ReadDataTo(7));
-        // assert_eq!(cache.data_listen_next(&d0, 4), ReadDataTo(7));
-        assert_eq!(cache.data_listen_next(&d0, 5), ReadDataTo(7));
-        assert_eq!(cache.data_listen_next(&d0, 6), ReadDataTo(7));
+        // This attempts to exercise all the various interesting edge cases of
+        // data_snapshot and data_listen_subscribe using the following sequence
+        // of events:
+        //
+        // - Registrations at: [2,8], [12,13]
+        // - Direct writes at: 1, 10
+        // - Writes via txns at: 4, 5, 7
+        let mut c = TxnsCacheState::new(ShardId::new(), 0, None);
+
+        // empty
+        assert_eq!(c.progress_exclusive, 0);
+        assert!(std::panic::catch_unwind(|| c.data_snapshot(d0, 0)).is_err());
+        assert_eq!(c.data_listen_next(&d0, 0), WaitForTxnsProgress);
+
+        // ts 0 (never registered)
+        testcase(&mut c, 0, d0, ds(None, 0, 1), ReadDataTo(1));
+
+        // ts 1 (direct write)
+        // - The cache knows everything < 2.
+        // - d0 is not registered in the cache.
+        // - We know the shard can't be written to via txn < 2.
+        // - So go read the shard normally up to 2.
+        testcase(&mut c, 1, d0, ds(None, 1, 2), ReadDataTo(2));
+
+        // ts 2 (register)
+        c.push_register(d0, 2, 1, 2);
+        testcase(&mut c, 2, d0, ds(None, 2, 3), EmitLogicalProgress(3));
+
+        // ts 3 (registered, not written)
+        testcase(&mut c, 3, d0, ds(None, 3, 4), EmitLogicalProgress(4));
+
+        // ts 4 (written via txns)
+        c.push_append(d0, vec![4], 4, 1);
+        testcase(&mut c, 4, d0, ds(Some(4), 4, 5), ReadDataTo(5));
+
+        // ts 5 (written via txns, write at preceding ts)
+        c.push_append(d0, vec![5], 5, 1);
+        testcase(&mut c, 5, d0, ds(Some(5), 5, 6), ReadDataTo(6));
+
+        // ts 6 (registered, not written, write at preceding ts)
+        testcase(&mut c, 6, d0, ds(Some(5), 6, 7), EmitLogicalProgress(7));
+
+        // ts 7 (written via txns, write at non-preceding ts)
+        c.push_append(d0, vec![7], 7, 1);
+        testcase(&mut c, 7, d0, ds(Some(7), 7, 8), ReadDataTo(8));
+
+        // ts 8 (forget). Forget guarantees that all writes are applied, so no
+        // latest_write. It also currently is inclusive, so we can emit logical
+        // progress, though we might want to make the interval half-open?
+        c.push_register(d0, 2, -1, 2);
```

Comment:
Good question. Seems like ideally it'd at least have one of the 2s be an 8. Less clear to me if we want to make both of them be 8. On one hand, it'd be nice to test our actual behavior. On the other, it feels a bit weird to be hardcoding the retraction bug in the test

Reply:
Ok, I've been thinking about this some more. I think for now both should be `8`. There's an invariant that `ts <= compact_ts` so if we change `ts` to `8`, we'll also need to change `compact_ts` to `8`. I think `ts` should be `8` because we definitely need to encode the timestamp of the forget itself to be able to reason about when the forget happened. How we encode the corresponding register timestamp is still an open question.

I've left a comment that includes a link to the forget consolidation issue so that when we fix that issue Nightly will start failing with invalid GH references.

Reply:
Makes sense to me!